pca_cca.go

```diff
@@ -0,0 +1,253 @@
+		return nil, nil, false
+	}
+
+	vecs = &mat64.Dense{}
+	vecs.VFromSVD(svd)
+	if n < d {
+		// Don't retain columns that are not valid direction vectors.
+		vecs.Clone(vecs.View(0, 0, d, n))
+	}
+	vars = svd.Values(nil)
+	var f float64
+	if weights == nil {
+		f = 1 / float64(n-1)
+	} else {
+		f = 1 / (floats.Sum(weights) - 1)
+	}
+	for i, v := range vars {
+		vars[i] = f * v * v
+	}
+	return vecs, vars, true
+}
+
+// CC is the result of a canonical correlation analysis. A CC should not be used
+// unless it has been returned by a successful call to CanonicalCorrelations.
+type CC struct {
+	// n is the number of observations used to construct the canonical
+	// correlations.
+	n int
+
+	x, y, c *mat64.SVD
+}
+
+// Corrs returns the canonical correlations in dest.
+func (c CC) Corrs(dest []float64) []float64 {
+	if c.c == nil {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	return c.c.Values(dest)
+}
+
+// Left returns the left-eigenvectors of the canonical correlation matrix if
+// spheredSpace is true, if spheredSpace is false it returns these eigenvectors
+// back-transformed to the original data space.
+func (c CC) Left(spheredSpace bool) (vecs *mat64.Dense) {
+	if c.c == nil || c.x == nil || c.n < 2 {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	var pv mat64.Dense
+	pv.UFromSVD(c.c)
+	if spheredSpace {
+		return &pv
+	}
+
+	var xv mat64.Dense
+	xs := c.x.Values(nil)
+	xv.VFromSVD(c.x)
+
+	scaleColsReciSqrt(&xv, xs)
+
+	pv.Product(&xv, xv.T(), &pv)
+	pv.Scale(math.Sqrt(float64(c.n-1)), &pv)
+	return &pv
+}
+
+// Right returns the right-eigenvectors of the canonical correlation matrix if
+// spheredSpace is true, if spheredSpace is false it returns these eigenvectors
+// back-transformed to the original data space.
+func (c CC) Right(spheredSpace bool) (vecs *mat64.Dense) {
+	if c.c == nil || c.y == nil || c.n < 2 {
+		panic("stat: canonical correlations missing or invalid")
+	}
+
+	var qv mat64.Dense
+	qv.VFromSVD(c.c)
+	if spheredSpace {
+		return &qv
+	}
+
+	var yv mat64.Dense
+	ys := c.y.Values(nil)
+	yv.VFromSVD(c.y)
+
+	scaleColsReciSqrt(&yv, ys)
+
+	qv.Product(&yv, yv.T(), &qv)
+	qv.Scale(math.Sqrt(float64(c.n-1)), &qv)
+	return &qv
+}
+
+// CanonicalCorrelations returns a CC which can provide the results of canonical
+// correlation analysis. The canonical correlation matrix is calculated using
+// the singular value decompositions of the centered input.
+//
+// CanonicalCorrelations will panic if the inputs x and y do not have the same
+// number of rows. Columns of x and y should be interpretable as two different
+// sets of measurements on the same observations.
+//
+// The slice weights is used to weight the observations. If weights is nil, each
+// weight is considered to have a value of one, otherwise the length of weights
+// must match the number of observations or CanonicalCorrelations will panic.
+//
+// Before describing canonical correlation analysis we first introduce some
```

Comment:
Something like this (which is incomplete and still not ideal):

```
// CanonicalCorrelations returns a CC which can provide the results of canonical
// correlation analysis. The canonical correlation matrix is calculated using
// the singular value decompositions of the centered input.
//
// Canonical correlation analysis finds linear combinations of pairs of random
// variables in a pair of joint dataset which have maximum correlation with each
// other. {{Explanation of sphering}}.
//
// When Xc and Yc denote the centered input data matrices x and y, Sx and Sy
// denote the sample covariance matrices within x and y respectively, and Sxy
// denotes the covariance matrix between x and y. The sphered data are
// Xc * Sx^(-1/2) and Yc * Sy^(-1/2) respectively, and the correlation matrix
// between the sphered data is the canonical correlation matrix,
// Sx^(-1/2) * Sxy * Sy^(-1/2). In cases where S^(-1/2) is ambiguous for some
// covariance matrix S, S^(-1/2) is taken to be E * D^(-1/2) * E' where S can
// be eigendecomposed as S = E * D * E'.
//
// The canonical correlations are the correlations between the corresponding
// pairs of canonical variables and can be obtained with c.Corrs(). Canonical
// variables can be obtained by projecting the sphered data into the left and
// right eigenvectors of the canonical correlation matrix, and these
// eigenvectors can be obtained with c.VectorsX(true) and c.VectorsY(true)
// respectively.
//
// The first pair of left and right eigenvectors of the canonical correlation
// matrix can be interpreted as directions into which the respective sphered
// data can be projected such that the correlation between the two projections
// is maximised. The second pair and onwards solve the same optimization but
// under the constraint that they are uncorrelated to previous projections.
//
// CanonicalCorrelations will panic if the inputs x and y do not have the same
// number of rows. Columns of x and y should be interpretable as two different
// sets of measurements on the same observations.
//
// The slice weights is used to weight the observations. If weights is nil, each
// weight is considered to have a value of one, otherwise the length of weights
// must match the number of observations or CanonicalCorrelations will panic.
//
// More details can be found at
// https://en.wikipedia.org/wiki/Canonical_correlation
// or in Chapter 3 of
// Koch, Inge. Analysis of multivariate and high-dimensional data.
// Vol. 32. Cambridge University Press, 2013. ISBN: 9780521887939
```